lxd/lxd: Validate https address for config key core.https_address

my5t3ry opened the pull request:
Fixes #8916 

Signed-off-by: Sascha Bast <sash@mischkonsum.org>

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
Hmm, I don't think this is correct.

`core.https_address` and `core.debug_address` don't quite take the same configuration as `cluster.https_address`.

`cluster.https_address` only support fixed addresses + port, it doesn't support things like `:8443`, `0.0.0.0:8443`, `[::]:8443` or relying on DNS names (like `localhost`, `localhost:8443`, ...). But both `core.https_address` and `core.debug_address` do support all of those and so can't share the same validator.

my5t3ry commented:
ok i was not aware of this. i added a own validator for core.https_address and core.debug_address now. it validates if the host is a valid IPV4,IPV6 address or a valid DNS name. But i encountert another problem while testing it. If i use a DNS name which can't be resolved lxd and systemd-resolve crashes and does so on restart until i reverted /var/lib/lxd/database/local.db. 
Is this a problem which can/should be addressed?
thanks : )

tomponline commented:
> ok sorry, i wasn't aware of this.
> 
> i added a own validator for core.https_address and core.debug_address now. it validates if the host is a valid IPV4,IPV6 address, :8443, 0.0.0.0:8443, [::]:8443, or a resolveable Hostname.
> I encountert a problem If i changed core.https_address or core.https_address to an unresolveable Hostname. It resulted in systemd-resolved doing something with 100% cpu utilization for ca 2 minutes and after it normalized lxd responded with.
> `Error: Put "http://unix.socket/1.0": EOF`
> Killing and restarting lxd resulted in the same systemd-resolved behavior and lxd got stuck while initializing. I wasn't able to get lxd initialized until i rolled back the config changes @/var/lib/lxd/database/local.db.
> 
> Calling net.Listner("tcp", address) with a unresolveable Hostname will never finish.
> 
> To prevent this case I added a check if the Hostname can be resolved to the validators.
> 
> : )

That sounds like a bug, can you give an example of a value passed to `net.Listener("tcp", address)` that would cause the crash/unresponsive behaviour?

my5t3ry commented:
sure,

net.Listener("tcp", "llocalhost:8443")
net.Listener("tcp", "kkernel.org:80")

my5t3ry commented:
i was wrong with guessing it never finishes it just takes a long time. i guess till some timeout occurs. and because it is repeated 10 times in cluster.go:ClusterUpdateAddress() or endpoints.go:NetworkUpdateAddress() it seemd only that it never finishes.

examples are:
net.Listen("tcp", "llocalhost:8443")
net.Listen("tcp", "kkernel.org:8443")

stgraber commented on lxd/node/config.go:
```diff
@@ -169,16 +169,56 @@ var ConfigSchema = config.Schema{
+		return nil // Deleting entry
+	}
+
+	host, _, err := net.SplitHostPort(value)
+	if err != nil {
+		return errors.Wrap(err, "Address not in form of <HOST>:<PORT>")
+	}
```
I thought we currently supported just `::` or any other address and it would then default to the usual port number (8443 or 8444 in this case).

stgraber commented on lxd/node/config.go:
```diff
@@ -169,16 +169,56 @@ var ConfigSchema = config.Schema{
+		return errors.Wrap(err, "Address not in form of <HOST>:<PORT>")
+	}
+
+	if shared.StringInSlice(host, []string{"::", "0.0.0.0"}) {
+		return nil
+	}
```
Those should be valid addresses and so pass ParseIP, no?

my5t3ry replied:
yes, you are right. i changed this.

tomponline commented:
jenkins: test this please

tomponline commented on lxd/node/config.go:
```diff
@@ -169,16 +168,53 @@ var ConfigSchema = config.Schema{
+		return nil
+	}
+
+	return fmt.Errorf("Only IPV4, IPV6 address, :<HOST>:<PORT>, or a valid and resolvable Hostname are valid")
```
Is that first colon in the error message a mistake, isn't it the same form as in `validateClusterHTTPSAddress`, i.e `<HOST>:<PORT>` not `:<HOST>:<PORT>`?

stgraber replied:
I guess it's either a typo or a mix of things. It's getting hard to cover everything that's supported in the error message though. What's supported is:
 - Wildcard addresses without a port (`0.0.0.0`, `::`, `[::]`)
 - Wildcard addresses with a port (`0.0.0.0:8443`, `[::]:8443`)
 - IPv4/IPv6 address without a port (`1.2.3.4`, `1234:4567`, `[1234::4567]`)
 - IPv4/IPv6 address with a port (`1.2.3.4:8443`, `[1234::4567]:8443`)
 - DNS record without a port (`blah`)
 - DNS record with a port (`blah:8443`)
 - Port only (`:8443`)

stgraber replied:
So we probably ought to check the validator against all of those ^ but the error itself should probably just be `Bad or unavailable listen address: %s'

tomponline replied:
Yeah makes sense, the inconsistency just stood out as a warning sign to me :)

tomponline commented on lxd/node/config.go:
```diff
@@ -169,16 +168,53 @@ var ConfigSchema = config.Schema{
+		return nil
+	}
+
+	return fmt.Errorf("Only valid IPV4, IPV6 aaddresses or resolvable hostnames are valid")
```
Why is this error different from the one in validateCoreHTTPSAddress, even though in this function higher up it has an error that says `errors.Wrap(err, "Address not in form of <HOST>:<PORT>")`? Seems inconsistent in messaging. Do we only accept an IP (and no port here)? If so, then the error above needs changing.

stgraber commented:
I've done some tweaks here, moving the logic into a single validator in `shared/validate` and porting the code over to that.
I also validated its behavior against the entire set of values provided above and it looks correct.

stgraber commented:
jenkins: test this please
